src/julia.jl

```diff
@@ -33,4 +33,56 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
     multireshape(parent, dimss...)
 end
 
+"""
+    pairwise(f::Function, a; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`. Set
+keyword `symmetric = true` to only calculate `f(x,y)` when `f(x,y) == f(y,x)`.
+"""
+function pairwise(f::Function, a; symmetric = false)
+    n = length(a)
+    first = f(a[1], a[1])
+    r = Matrix{typeof(first)}(n, n)
+    pairwise!(f, a, r; symmetric = symmetric)
+    r
+end
+
+"""
+    pairwise!(f::Function, a, r)
+
+Apply the function `f` to all pairwise combinations of elements from `a` and
+storing the results in `r`. Setting keyword `symmetric = true` will only calculate
+each variable combination once for symmetric functions. This is useful if `f` is
+expensive.
+"""
+function pairwise!(f::Function, a, r::AbstractMatrix; symmetric = false)
+    n = length(a)
+    size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r ($(size(r)), should be $((n, n)))"))
+    if symmetric
+        pairwise_symmetric!(f, a, r, n)
+    else
+        pairwise_default!(f, a, r, n)
+    end
+end
+
+function pairwise_default!(f, a, r, n)
+    @inbounds for j = 1:n
```

Comment:
(However, you wouldn't want to do this in the case where the user called `pairwise!` ... in that case I would just throw an error.)

Of course, it may not be worthwhile to go to this trouble here.  For now, you could just do `r[i,j] = f(a[i],a[j])` and it will throw an error if the types are incompatible